### PR TITLE
[DROOLS-1536] allow to cache compiler subcomponents into the Knowledg…

### PR DESCRIPTION
**Pull request by mariofusco**
…eBuilderImpl

### CODE REVIEW

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/assembler/DMNAssemblerService.java [hunk 1]
```diff
@@ -56,13 +56,13 @@ public ResourceType getResourceType() {
     @Override
     public void addResource(KnowledgeBuilder kbuilder, Resource resource, ResourceType type, ResourceConfiguration configuration)
             throws Exception {
-        DMNCompiler dmnCompiler = getCompiler(kbuilder);
+        KnowledgeBuilderImpl kbuilderImpl = (KnowledgeBuilderImpl) kbuilder;
+        DMNCompiler dmnCompiler = kbuilderImpl.getCachedOrCreate( DMN_COMPILER_CACHE_KEY, () -> getCompiler( kbuilderImpl ) );
```
Comment (tarilabs) [inline kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/assembler/DMNAssemblerService.java:60]: good for me, thanks! cc/ @etirelli

**Review by tarilabs — APPROVED** [review]
+1

**Comment by etirelli** [comment]
@mariofusco this is not how I would have done it, but looks good to me! :)

**Comment by etirelli** [comment]
@mariofusco BTW, thank you for doing this so quick... it is *really* important for the DMN compiler.

**Comment by mariofusco** [comment]
@etirelli What did you have in mind? Let me know if you have any suggestion to improve this.

**Comment by etirelli** [comment]
@mariofusco I think this should work fine, as I believe the kbuilder is scoped per kbase in the kjar, right? My usual pattern would be to add a "createContext()" method to the Assembler interface and then pass back the created context as parameter to each method call. Your solution is simpler and works fine as long as we guarantee the kbuilder lifecycle.

**Comment by mariofusco** [comment]
@etirelli I was doing what you're also suggesting, but yes then I realized that we wanted something with same lifecycle of the kbuilder so I could just add the cache directly inside it. This also allowed to avoid the need of an additional argument in the assembler interface.

**Pull request merged**